### PR TITLE
Speed up tests by running the biggest ones first

### PR DESCRIPTION
**Pull request by kwiberg**
And default to one thread per CPU.

### CODE REVIEW

**Comment by kwiberg** [comment]
Updated the pull request. (Patch 2/3: Now store just the latest time for each test in a single file. Patch 3/3: Eliminated the named function.)

File: gtest-parallel [hunk 1]
```diff
@@ -82,6 +87,45 @@ class RawFormat:
   def end(self):
     pass
 
+# Record of test runtimes. Has built-in locking.
+class TestTimes(object):
+  def __init__(self, save_file):
+    "Create new object seeded with saved test times from the given file."
+    self.__old_times = {}  # (test binary, test name) -> runtime in seconds
```
Comment (pbos) [inline gtest-parallel]: Do you need both old_times and new_times? get_test_time is only called before tests are running, right?

Also, imo remove the __ prefixes, unless you can point me to somewhere saying they're Google style.

File: gtest-parallel [hunk 2]
```diff
@@ -82,6 +87,45 @@ class RawFormat:
...
+    self.__new_times = {}  # (test binary, test name) -> runtime in seconds
+
+    # Protects __new_times; __old_times doesn't need locking since
+    # it's read-only after construction.
+    self.__lock = threading.Lock()
+
+    try:
+      with gzip.GzipFile(save_file, "rb") as f:
+        self.__old_times = cPickle.load(f)
```
Comment (pbos) [inline gtest-parallel]: Is there any chance that you can validate that old_times is now a (test binary, test name) -> runtime in seconds? Or at least (string, string) -> float so we don't get any runtime errors if we change file format.

File: gtest-parallel (hunk 1 above)
Comment (kwiberg) [inline gtest-parallel]: I think you're right regarding old_times and new_times. I'll combine them.

Prefixing with __ isn't about style, it's how member variables are made private in Python. See e.g. https://docs.python.org/2/tutorial/classes.html#private-variables-and-class-local-references.

File: gtest-parallel (hunk 2 above)
Comment (kwiberg) [inline gtest-parallel]: That's a good idea. Will do.

File: gtest-parallel [hunk 3]
```diff
@@ -175,11 +236,12 @@ def run_job((command, job_id, test)):
     logger.log(str(job_id) + '> ' + line.rstrip())
 
   code = sub.wait()
-  runtime_ms = int(1000 * (time.time() - begin))
-  logger.log(str(job_id) + ': EXIT ' + str(code) + ' ' + str(runtime_ms))
+  runtime = time.time() - begin
```
Comment (pbos) [inline gtest-parallel]: runtime_ms so we can be consistent, any sub-ms precision doesn't matter either way

**Pull request merged**